File: app/src/main/java/com/sarnavsky/pasz/nighlight2/Fragments/MainFragment.java
package com.sarnavsky.pasz.nighlight2.Fragments;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.util.Log;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.view.animation.LinearInterpolator;
import android.view.animation.RotateAnimation;
import android.view.animation.ScaleAnimation;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.RequestConfiguration;
import com.google.android.gms.ads.initialization.AdapterStatus;
import com.google.android.gms.ads.initialization.InitializationStatus;
import com.google.android.gms.ads.initialization.OnInitializationCompleteListener;
import com.sarnavsky.pasz.nighlight2.Adapters.Nightlighters;
import com.sarnavsky.pasz.nighlight2.Adapters.RecyclerViewAdapter;
import com.sarnavsky.pasz.nighlight2.Fabrica.MyObjects;
import com.sarnavsky.pasz.nighlight2.Interfaces.ChangeColors;
import com.sarnavsky.pasz.nighlight2.Interfaces.OpenColorFragment;
import com.sarnavsky.pasz.nighlight2.MainActivity;
import com.sarnavsky.pasz.nighlight2.Objects.MenuButton;
import com.sarnavsky.pasz.nighlight2.Objects.Nightlighter;
import com.sarnavsky.pasz.nighlight2.R;

import java.util.ArrayList;
import java.util.Map;


public class MainFragment extends Fragment {

    //MyADD
    public static AdView mAdView;
    AdRequest adRequest;

//CAS
    // public static CASBannerView bannerView;

    FrameLayout lock_frame;
    FragmentManager fm;
    ViewPager2 pager;
    CountDownTimer cdt;
    CountDownTimer globalTimer;
    ConstraintLayout mainBg;
    RecyclerViewAdapter adapter;
    RecyclerView rv;
    Context ctx;
    MyObjects menuItems;
    public TextView bottom_text;

    boolean showed;
    boolean chekAnim = false;
    boolean show = true;
    boolean timerStatus = false;
    boolean chekMenu = true;

    int currentBgColor = 0;
    int currentBgImage = 0;
    int currentNLColor = 0;
    int brights = 0;

    String[] colors;
    String[] bgColors;
    String[] bgNlColors;
    String[] menuColors;

    ImageView underImg;
    ImageView lock_button;
    ImageView settings_button;
    ImageView animateBg;

    // View objects to display banners:
    RelativeLayout topBannerView;
    RelativeLayout bottomBannerView;
    // Buttons to show the banners:
    Button showTopBannerButton;
    Button showBottomBannerButton;
    // Buttons to destroy the banners:
    Button hideTopBannerButton;
    Button hideBottomBannerButton;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.main_fragment, container, false);
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onViewCreated(final View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ctx = view.getContext();
        FragmentActivity main = getActivity();
        fm = getParentFragmentManager();



        //chekSubscription
//        Log.i("TESTSUB", " chekSubscription");
//        PurchasesUpdatedListener purchasesUpdatedListener = new PurchasesUpdatedListener() {
//            @Override
//            public void onPurchasesUpdated(@NonNull BillingResult billingResult, @Nullable List<Purchase> list) {
//            }
//        };
//        BillingClient bc = BillingClient.newBuilder(getContext()).enablePendingPurchases().setListener(purchasesUpdatedListener).build();
//        bc.startConnection(new BillingClientStateListener() {
//            @Override
//            public void onBillingServiceDisconnected() {
//            }
//
//            @Override
//            public void onBillingSetupFinished(@NonNull BillingResult billingResult) {
//
//                Log.i("TESTSUB", billingResult + " billingResult");
//
//                QueryPurchasesParams queryPurchasesParams = QueryPurchasesParams
//                        .newBuilder()
//                        .setProductType(BillingClient.ProductType.SUBS)
//                        .build();
//
//                bc.queryPurchasesAsync(queryPurchasesParams, new PurchasesResponseListener() {
//                    @Override
//                    public void onQueryPurchasesResponse(@NonNull BillingResult billingResult, @NonNull List<Purchase> list) {
//
//
//                        Log.i("TESTSUB", list.size() + " TestMy");
//
//                        getActivity().runOnUiThread(new Runnable() {
//
//                            @Override
//                            public void run() {
//
//                                if (billingResult.getResponseCode() == 0) {
//                                    if (list.size() > 0) {
//                                        mAdView.setVisibility(View.GONE);
//                                    } else {
//                                        mAdView.loadAd(adRequest);
//                                        getParentFragmentManager().beginTransaction().add(R.id.container, new Subscription(), "subscription").commit();
//                                    }
//                                }
//
//                            }
//                        });
//
//
//                    }
//                });
//            }
 //       });


        //end chekSubscription

        //Start Global Itmer. Make all invisible exept nightlighter.
        startGlobalTimer();

        //UnityAds.SetC("privacy.consent", true);

        //MyADMOB ADD
        //findAdsView
        mAdView = view.findViewById(R.id.adView);



        //MyADMOB ADD
        //Use ADDS
        setAdsSetting();
        //MyADMOB ADD
        adRequest = new AdRequest.Builder().build();


        int addCounter = ((MainActivity) getActivity()).getSettings();
        if (addCounter > 0) {
            //MyADMOB ADD
            mAdView.setVisibility(View.GONE);

        } else {
            //MyADMOB ADD
            mAdView.loadAd(adRequest);
        }


        mainBg = (ConstraintLayout) view.findViewById(R.id.mainBg);
        menuItems = new MyObjects();
        lock_button = view.findViewById(R.id.lock_button);
        settings_button = view.findViewById(R.id.settings_button);

        settings_button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                SettingsFragment settingsFragment = (SettingsFragment) getParentFragmentManager().findFragmentByTag("SETTINGS_FRAGMENT");
                if (settingsFragment == null) {
                    getParentFragmentManager()
                            .beginTransaction()
                            .setCustomAnimations(R.anim.from_left, R.anim.disepire)
                            .replace(R.id.mainContainer, new SettingsFragment(), "SETTINGS_FRAGMENT")
                            .commit();
                }

            }
        });

        lock_button = (ImageView) view.findViewById(R.id.lock_button);
        animateBg = (ImageView) view.findViewById(R.id.animateBg);
        lock_frame = view.findViewById(R.id.lock_frame);

        lock_frame.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                showButtons();
                startGlobalTimer();
                return false;
            }
        });

        rv = (RecyclerView) view.findViewById(R.id.rv);
        LinearLayoutManager llm = new LinearLayoutManager(view.getContext(), LinearLayoutManager.HORIZONTAL, false);
        rv.setLayoutManager(llm);
        Resources res = getResources();
        colors = res.getStringArray(R.array.myColors);
        bgColors = res.getStringArray(R.array.bgColors);
        bgNlColors = res.getStringArray(R.array.bgNlColors);
        menuColors = res.getStringArray(R.array.menuColors);

        ArrayList<Nightlighter> arrayList = new ArrayList<>();
        MyObjects myMenuItems = new MyObjects();
        arrayList.addAll(myMenuItems.getNightlighters());

        pager = view.findViewById(R.id.pager);
        pager.setAdapter(new Nightlighters(arrayList, bgNlColors));
        pager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                super.onPageSelected(position);
                bottom_text = view.findViewById(R.id.bottom_text);
                bottom_text.setText(arrayList.get(pager.getCurrentItem()).getName());
            }

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                super.onPageScrolled(position, positionOffset, positionOffsetPixels);
                underImg = pager.findViewById(R.id.underImg);
            }
        });

        lock_button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                lockButton();
            }
        });

        openMenu(menuItems.getMenuButtons(colors));
    }

    public void openMenu(ArrayList<MenuButton> menuButtons) {

        adapter = new RecyclerViewAdapter(menuButtons, menuColors);
        adapter.MyOnclick(new ChangeColors() {
            @Override
            public void onclick(int button) {
                switch (button) {
                    case 1:
                        ListFragment listFragment = (ListFragment) getChildFragmentManager().findFragmentByTag("LIST_FRAGMENT");

                        if (listFragment == null) {
                            getParentFragmentManager()
                                    .beginTransaction()
                                    .setCustomAnimations(R.anim.from_bottom, R.anim.disepire)
                                    .replace(R.id.mainContainer, new ListFragment(), "LIST_FRAGMENT")
                                    .commit();
                        }
                        break;
                    case 2:
                        showBgColorMenu();
                        break;
                    case 3:
                        changeNLColor();
                        break;
                    case 4:
                        startAnimation();

                        break;
                    case 6:
                        changeBgColor();
                        break;
                    case 5:
                        TimerFragment timer_fragment = (TimerFragment) getChildFragmentManager().findFragmentByTag("TIMER_FRAGMENT");

                        if (timer_fragment == null) {
                            getParentFragmentManager()
                                    .beginTransaction()
                                    .setCustomAnimations(R.anim.from_bottom, R.anim.disepire)
                                    .replace(R.id.mainContainer, new TimerFragment(), "TIMER_FRAGMENT")
                                    .commit();
                        }

                        break;
                    case 7:
                        changeBrighest();
                        break;
                }
            }
        });
        adapter.MyOnLongclick(new OpenColorFragment() {
            @Override
            public void onclick(int button) {
                switch (button) {
                    case 3:
                        getParentFragmentManager()
                                .beginTransaction()
                                .replace(R.id.mainContainer, ColorPicker.init(0), "ColorPicker")
                                .commit();
                        break;
                    case 2:
                        getParentFragmentManager()
                                .beginTransaction()
                                .replace(R.id.mainContainer, ColorPicker.init(1), "ColorPicker")
                                .commit();
                        break;
                }
            }
        });
        rv.setAdapter(adapter);
    }

    private void showBgColorMenu() {

        currentBgColor++;
        if (currentBgColor == bgColors.length) {
            currentBgColor = 0;
        }

        mainBg.setBackgroundColor(Color.parseColor(bgColors[currentBgColor]));
    }

    public void changeBrighest() {
        WindowManager.LayoutParams layout = getActivity().getWindow().getAttributes();

        switch (brights) {
            case 0:
                layout.screenBrightness = 0.1F;
                brights++;
                break;
            case 1:
                layout.screenBrightness = 0.5F;
                brights++;
                break;
            case 2:
                layout.screenBrightness = 1F;

                brights = 0;
                break;
        }

        getActivity().getWindow().setAttributes(layout);
    }

    public void startAnimation() {

        animateBg.setScaleType(ImageView.ScaleType.FIT_CENTER);

        if (!chekAnim) {
            RotateAnimation rotate = new RotateAnimation(0f, 360f,
                    Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f); //8
            rotate.setDuration(100000);
            rotate.setRepeatCount(Animation.INFINITE);
            rotate.setInterpolator(new LinearInterpolator());

            AnimationSet set = new AnimationSet(false); //10
            set.addAnimation(rotate);
            animateBg.startAnimation(set);
            chekAnim = true;

            //float mySacele = (float) getResources().getDimension(R.dimen.scale);
           // Log.i("DIMEN", mySacele+"ddd");

            TypedValue outValue = new TypedValue();
            getResources().getValue(R.dimen.scale, outValue, true);
            float value = outValue.getFloat();
            Log.i("DSSD",value+"  d");

            ScaleAnimation scale = new ScaleAnimation(1f, value, 1f, value,
                    Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
            scale.setDuration(1000);
            animateBg.startAnimation(scale);

            set.addAnimation(scale);
            animateBg.startAnimation(set);


        } else {
            animateBg.clearAnimation();
            animateBg.setScaleType(ImageView.ScaleType.CENTER_CROP);
            chekAnim = false;
        }

    }

    public void lockButton() {

        int showAdd = ((MainActivity) getActivity()).getSettings();
        if (chekMenu) {
            lock_frame.setClickable(true);
            openMenu(menuItems.getMenuButtons(colors));
            rv.setVisibility(View.INVISIBLE);
            bottom_text.setVisibility(View.GONE);
            settings_button.setVisibility(View.GONE);

            int frCount = getChildFragmentManager().getFragments().size();
            //Toast.makeText(ctx, frCount+" ddd", Toast.LENGTH_SHORT).show();
            if (frCount > 0) {
                for (int i = 0; i < frCount; i++) {
                    Fragment fragment = getChildFragmentManager().getFragments().get(i);
                    getChildFragmentManager().beginTransaction().remove(fragment).commit();
                }
            }

            lock_button.setImageResource(R.drawable.ic_lock);

            if (showAdd > 0) {

            } else {
                //MyADMOB ADD
                mAdView.setVisibility(View.INVISIBLE);
            }


            lock_frame.setClickable(true);
            chekMenu = false;
            show = false;
        } else {

            lock_frame.setClickable(false);
            rv.setVisibility(View.VISIBLE);
            bottom_text.setVisibility(View.VISIBLE);
            lock_button.setImageResource(R.drawable.ic_unlock);

            if (showAdd > 0) {

            } else {
                //MyADMOB ADD
                mAdView.setVisibility(View.VISIBLE);
            }


            settings_button.setVisibility(View.VISIBLE);
            lock_frame.setClickable(false);
            chekMenu = true;
            show = true;
        }

    }

    public void startTimer(int hours, int minutes) {
        bottom_text.setVisibility(View.VISIBLE);
        int mySeconds = (((hours * 60 * 60) + (60 * minutes)) * 1000);
        closeApp(mySeconds);
    }

    public void closeApp(int mySeconds) {

        if (cdt != null) {
            timerStatus = false;
            cdt.cancel();
            cdt = null;
        }
        if (mySeconds > 0) {
            timerStatus = true;
            cdt = new CountDownTimer(mySeconds, 1000) {
                @SuppressLint("DefaultLocale")
                @Override
                public void onTick(long l) {
                    bottom_text.setText(String.format("%02d:%02d:%02d", (l / 1000) / 3600, ((l / 1000) % 3600) / 60, (l / 1000) % 60));
                }

                @Override
                public void onFinish() {

                    if (cdt != null) {
                        cdt.cancel();
                    }

                    if (globalTimer != null) {
                        globalTimer.cancel();
                    }

                    ((MainActivity) getActivity()).finishMedia();

                    Log.i("FINISH", "App is OFF");
                    getActivity().finish();
                }
            };
            cdt.start();
        } else {
            bottom_text.setText("");
            bottom_text.setVisibility(View.INVISIBLE);
        }
    }

    private void setAdsSetting() {
        // Initialize the Mobile Ads SDK.
        MobileAds.initialize(getContext(), new OnInitializationCompleteListener() {
            @Override
            public void onInitializationComplete(InitializationStatus initializationStatus) {
                Map<String, AdapterStatus> statusMap = initializationStatus.getAdapterStatusMap();
                for (String adapterClass : statusMap.keySet()) {
                    AdapterStatus status = statusMap.get(adapterClass);
                    Log.d("MyApp", String.format(
                            "Adapter name: %s, Description: %s, Latency: %d",
                            adapterClass, status.getDescription(), status.getLatency()));
                }

                // Start loading ads here...
            }
        });

        RequestConfiguration requestConfiguration = MobileAds.getRequestConfiguration()
                .toBuilder()
                .setTagForChildDirectedTreatment(RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_TRUE)
                .build();
        MobileAds.setRequestConfiguration(requestConfiguration);

    }

    private void changeBgColor() {

        currentBgColor++;
        currentBgImage++;

        if (currentBgImage >= menuItems.getBgArray().length) {
            currentBgImage = 0;
            currentBgColor = 0;
        }

        if (currentBgColor >= bgColors.length) {
            currentBgColor = 0;
        }

        mainBg.setBackgroundColor(Color.parseColor(bgColors[currentBgColor]));
        animateBg.setImageResource(menuItems.getBgArray()[currentBgImage]);
    }

    private void changeNLColor() {

        currentNLColor++;

        if (currentNLColor >= bgNlColors.length) {
            currentNLColor = 0;
        }

        underImg.setColorFilter(Color.parseColor(bgNlColors[currentNLColor]));
    }

    private void startGlobalTimer() {
        if (globalTimer != null) {
            globalTimer.start();
        } else {
            globalTimer = new CountDownTimer(5000, 1000) {
                @Override
                public void onTick(long l) {

                }

                @Override
                public void onFinish() {

                    if (lock_button != null) {
                        lock_button.setVisibility(View.INVISIBLE);
                    }
                    if (bottom_text != null) {
                        bottom_text.setVisibility(View.INVISIBLE);
                    }
                    if (settings_button != null) {
                        settings_button.setVisibility(View.INVISIBLE);
                    }
                    if (rv != null) {
                        rv.setVisibility(View.INVISIBLE);
                    }
                }
            }.start();
        }
    }

    private void showButtons() {
        if (chekMenu) {
            lock_button.setVisibility(View.VISIBLE);
            bottom_text.setVisibility(View.VISIBLE);
            settings_button.setVisibility(View.VISIBLE);
            rv.setVisibility(View.VISIBLE);
        } else {
            lock_button.setVisibility(View.VISIBLE);
            bottom_text.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/com/sarnavsky/pasz/nighlight2/fragments/MainFragment.kt
package com.sarnavsky.pasz.nighlight2.fragments

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.os.CountDownTimer
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationSet
import android.view.animation.LinearInterpolator
import android.view.animation.RotateAnimation
import android.view.animation.ScaleAnimation
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.viewpager2.widget.ViewPager2
import com.daimajia.androidanimations.library.YoYo
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.RequestConfiguration
import com.jaredrummler.android.colorpicker.ColorPickerDialog
import com.jaredrummler.android.colorpicker.ColorPickerDialogListener
import com.sarnavsky.pasz.nighlight2.MainActivity
import com.sarnavsky.pasz.nighlight2.R
import com.sarnavsky.pasz.nighlight2.SettingsViewModel
import com.sarnavsky.pasz.nighlight2.adapters.MainMenuAdapter
import com.sarnavsky.pasz.nighlight2.adapters.NightlightersAdapter
import com.sarnavsky.pasz.nighlight2.data.db.entity.Settings
import com.sarnavsky.pasz.nighlight2.data.db.entity.Timer
import com.sarnavsky.pasz.nighlight2.databinding.MainFragmentBinding
import com.sarnavsky.pasz.nighlight2.util.ANIMATION_BUTTON
import com.sarnavsky.pasz.nighlight2.util.ANIMATION_TYPE_BUTTON
import com.sarnavsky.pasz.nighlight2.util.BG_COLOR_BUTTON
import com.sarnavsky.pasz.nighlight2.util.BRIGHTS_BUTTON
import com.sarnavsky.pasz.nighlight2.util.NL_COLOR_BUTTON
import com.sarnavsky.pasz.nighlight2.util.NightlightHelper
import com.sarnavsky.pasz.nighlight2.util.SOUNDS_BUTTON
import com.sarnavsky.pasz.nighlight2.util.TIMER_BUTTON
import org.koin.androidx.viewmodel.ext.android.activityViewModel
import java.util.Random

class MainFragment : Fragment() {



    lateinit var binding: MainFragmentBinding
    private val viewModel: SettingsViewModel by activityViewModel()

    private lateinit var mySetting: Settings
    private lateinit var myTimer: Timer

    var cdt: CountDownTimer? = null

    private lateinit var bgColors: Array<String>
    private lateinit var bgNlColors: Array<String>

    private var currentBgColor = 0
    private var currentBgImage = 0
    private var currentNLColor = 0
    private var brights = 0
    private var timerIsLoaded = false
    private var checkMenu = true
    private var show = true
    private var checkAnim = false
    private var timerStatus = false

    private val mainMenuAdapter = MainMenuAdapter { menuItem, longClick ->
        if (longClick) {
            when (menuItem.button) {
                BG_COLOR_BUTTON -> showColorPicker(BG_COLOR_BUTTON)
                NL_COLOR_BUTTON -> showColorPicker(NL_COLOR_BUTTON)
            }
        } else {
            when (menuItem.button) {
                SOUNDS_BUTTON -> (requireActivity() as MainActivity).openFragment(MusicListFragment())
                BG_COLOR_BUTTON -> changeBackgroundColor()
                ANIMATION_BUTTON -> startAnimation()
                TIMER_BUTTON -> {
                    closeApp(0)
                    (requireActivity() as MainActivity).openFragment(TimerFragment())
                }

                NL_COLOR_BUTTON -> {
                    currentNLColor++
                    if (currentNLColor >= bgNlColors.size) {
                        currentNLColor = 0
                    }

                    val color = Color.parseColor(bgNlColors[currentNLColor])
                    changeNLColor(color)
                }

                ANIMATION_TYPE_BUTTON -> changeAnimationType()
                BRIGHTS_BUTTON -> changeBrightest()
            }
        }
    }
    private val nightlightersAdapter = NightlightersAdapter {

        val techniques = NightlightHelper.getTechniquesArray()
        val random = Random()
        val i = random.nextInt(techniques.size)

        YoYo.with(techniques[i])
            .duration(700)
            .playOn(binding.pager)

    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = MainFragmentBinding
            .inflate(inflater, container, false)
        return binding.root
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initArrays()
        initAdapter()
        initListener()
        observer()
        initView()

        initAds()


    }

    private fun initAdapter() {
        binding.rv.adapter = mainMenuAdapter
        val listMenu = NightlightHelper.getMenuButtons(resources)
        mainMenuAdapter.list.submitList(listMenu)

        binding.pager.adapter = nightlightersAdapter
        val listNightlighters = NightlightHelper.getNightlighters()
        nightlightersAdapter.list.submitList(listNightlighters)

    }

    private fun initAds() {
        MobileAds.initialize(
            requireContext()
        ) { initializationStatus ->
            val statusMap = initializationStatus.adapterStatusMap
            for (adapterClass in statusMap.keys) {
                val status = statusMap[adapterClass]
                Log.d(
                    "MyApp", String.format(
                        "Adapter name: %s, Description: %s, Latency: %d",
                        adapterClass, status?.description, status?.latency
                    )
                )
            }

        }
        val requestConfiguration = MobileAds.getRequestConfiguration()
            .toBuilder()
            .setTagForChildDirectedTreatment(RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_TRUE)
            .build()
        MobileAds.setRequestConfiguration(requestConfiguration)



        val adRequest = AdRequest.Builder().build()
        binding.adView.loadAd(adRequest)

        binding.adView.adListener = object: AdListener() {
            override fun onAdClicked() {
                // Code to be executed when the user clicks on an ad.
                Log.i("ADSMY", "Code to be executed when the user clicks on an ad.")
            }

            override fun onAdClosed() {
                // Code to be executed when the user is about to return
                // to the app after tapping on an ad.
                Log.i("ADSMY", " Code to be executed when the user is about to return")
            }

            override fun onAdFailedToLoad(adError : LoadAdError) {
                // Code to be executed when an ad request fails.
                Log.i("ADSMY", "Code to be executed when an ad request fails.")
            }

            override fun onAdImpression() {
                // Code to be executed when an impression is recorded
                // for an ad.
                Log.i("ADSMY", "Code to be executed when an impression is recorded")
            }

            override fun onAdLoaded() {
                // Code to be executed when an ad finishes loading.
                Log.i("ADSMY", "Code to be executed when an ad finishes loading.")
            }

            override fun onAdOpened() {
                // Code to be executed when an ad opens an overlay that
                // covers the screen.
                Log.i("ADSMY", "Code to be executed when an ad opens an overlay that")
            }
        }
    }

    private fun showButtons() {
        if (checkMenu) {
            binding.lockButton.visibility = View.VISIBLE
            binding.bottomText.visibility = View.VISIBLE
            binding.settingsButton.visibility = View.VISIBLE
            binding.rv.visibility = View.VISIBLE
        } else {
            binding.lockButton.visibility = View.VISIBLE
            binding.bottomText.visibility = View.VISIBLE
        }
    }

    private fun lockButton() {
        val showAdd = (activity as MainActivity?)?.getSettings()
        if (checkMenu) {
            binding.lockFrame.isClickable = true
            // openMenu(NightlightHelper.getMenuButtons(colors))
            binding.rv.visibility = View.INVISIBLE
            binding.bottomText.visibility = View.GONE
            binding.settingsButton.visibility = View.GONE
            val frCount = childFragmentManager.fragments.size
            //Toast.makeText(ctx, frCount+" ddd", Toast.LENGTH_SHORT).show();
            if (frCount > 0) {
                for (i in 0 until frCount) {
                    val fragment = childFragmentManager.fragments[i]
                    childFragmentManager.beginTransaction().remove(fragment).commit()
                }
            }
            binding.lockButton.setImageResource(R.drawable.ic_lock)
            if (showAdd != null) {
                if (showAdd < 0) {
                    binding.adView.visibility = View.INVISIBLE
                }
            }
            binding.lockFrame.isClickable = true
            checkMenu = false
            show = false
        } else {
            binding.lockFrame.isClickable = false
            binding.rv.visibility = View.VISIBLE
            binding.bottomText.visibility = View.VISIBLE
            binding.lockButton.setImageResource(R.drawable.ic_unlock)
            if (showAdd != null) {
                if (showAdd < 0) {
                    binding.adView.visibility = View.VISIBLE
                }
            }
            binding.settingsButton.visibility = View.VISIBLE
            binding.lockFrame.isClickable = false
            checkMenu = true
            show = true
        }
    }

    private fun changeBackgroundColor() {
        currentBgColor++
        if (currentBgColor == bgColors.size) {
            currentBgColor = 0
        }
        binding.mainBg.setBackgroundColor(Color.parseColor(bgColors[currentBgColor]))
        mySetting.backgroundColor = Color.parseColor(bgColors[currentBgColor])
        //viewModel.updateSettings(mySetting)

    }

    private fun changeNLColor(color: Int) {
        nightlightersAdapter.updateImageColorsWithColor(color)
        mySetting.nightlightColor = color
    }

    private fun startAnimation() {
        binding.animateBg.scaleType = ImageView.ScaleType.FIT_CENTER
        if (!checkAnim) {
            val rotate = RotateAnimation(
                0f, 360f,
                Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f
            ) //8
            rotate.duration = 100000
            rotate.repeatCount = Animation.INFINITE
            rotate.interpolator = LinearInterpolator()
            val set = AnimationSet(false) //10
            set.addAnimation(rotate)
            binding.animateBg.startAnimation(set)
            checkAnim = true

            val outValue = TypedValue()
            resources.getValue(R.dimen.scale, outValue, true)
            val value = outValue.float
            val scale = ScaleAnimation(
                1f, value, 1f, value,
                Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f
            )
            scale.duration = 1000
            binding.animateBg.startAnimation(scale)
            set.addAnimation(scale)
            binding.animateBg.startAnimation(set)
            mySetting.animationStatus = true
        } else {
            binding.animateBg.clearAnimation()
            binding.animateBg.scaleType = ImageView.ScaleType.CENTER_CROP
            checkAnim = false
            mySetting.animationStatus = false
        }
    }

    private fun changeAnimationType() {
        currentBgColor++
        currentBgImage++
        if (currentBgImage >= NightlightHelper.getBgArray().size) {
            currentBgImage = 0
            currentBgColor = 0
        }
        if (currentBgColor >= bgColors.size) {
            currentBgColor = 0
        }

        binding.animateBg.setImageResource(NightlightHelper.getBgArray()[currentBgImage])
        mySetting.animationType = NightlightHelper.getBgArray()[currentBgImage]
    }

    private fun changeBrightest() {
        val layout = activity?.window?.attributes
        when (brights) {
            0 -> {
                layout?.screenBrightness = 0.1f
                brights++
            }

            1 -> {
                layout?.screenBrightness = 0.5f
                brights++
            }

            2 -> {
                layout?.screenBrightness = 1f
                brights = 0
            }
        }
        activity?.window?.attributes = layout
    }

    private fun closeApp(mySeconds: Int) {
        if (cdt != null) {
            timerStatus = false
            cdt?.cancel()
            cdt = null
        }
        if (mySeconds > 0) {
            timerStatus = true
            cdt = object : CountDownTimer(mySeconds.toLong(), 1000) {
                @SuppressLint("DefaultLocale")
                override fun onTick(l: Long) {
                    binding.bottomText.text = String.format(
                        "%02d:%02d:%02d",
                        l / 1000 / 3600,
                        l / 1000 % 3600 / 60,
                        l / 1000 % 60
                    )
                }

                override fun onFinish() {
                    if (cdt != null) {
                        cdt?.cancel()
                    }
                    Log.i("FINISH", "App is OFF")
                    activity?.finish()
                }

            }
            cdt?.start()

        } else {
            binding.bottomText.text = ""
            binding.bottomText.visibility = View.INVISIBLE
        }
    }

    private fun observer() {

        viewModel.getSettings()
        viewModel.getTimer()

        viewModel.timerLiveData.observe(viewLifecycleOwner){
            if(it == null){
                viewModel.insertTimer()
            }else{
                myTimer = it
                timerIsLoaded = true
                if (it.timerStatus) {
                    binding.bottomText.visibility = View.VISIBLE
                    closeApp(it.timerDuration)
                }
            }

        }

        viewModel.settingsLiveData.observe(viewLifecycleOwner) {
            if (it == null) {
                viewModel.insertItem()
            } else {
                mySetting = it


                binding.mainBg.setBackgroundColor(it.backgroundColor)
                changeNLColor(it.nightlightColor)
                binding.pager.currentItem = it.currentNightlight

                binding.animateBg
                    .setImageResource(it.animationType)

                if (it.animationStatus) {
                    startAnimation()
                }

            }
        }
    }

    private fun showColorPicker(type: Int) {
        val colorPickerDialog = ColorPickerDialog
            .newBuilder()
            .setSelectedButtonText(R.string.selected_button)
            .setDialogTitle(R.string.dialog_title)
            .setCustomButtonText(R.string.custom_button)
            .setPresetsButtonText(R.string.presets_button)
            .setColor(Color.RED)
            .create()
        colorPickerDialog.show(parentFragmentManager, "")

        colorPickerDialog.setColorPickerDialogListener(
            object : ColorPickerDialogListener {
                override fun onColorSelected(dialogId: Int, color: Int) {

                    when (type) {
                        BG_COLOR_BUTTON -> {
                            mySetting.backgroundColor = color
                            binding.mainBg.setBackgroundColor(color)
                        }

                        NL_COLOR_BUTTON -> {
                            mySetting.nightlightColor = color
                            changeNLColor(color)
                        }
                    }
                }

                override fun onDialogDismissed(dialogId: Int) {
                }
            })
    }

    override fun onStop() {
        super.onStop()
        mySetting.currentNightlight = binding.pager.currentItem
        myTimer.timerStatus = false
        viewModel.updateSettings(mySetting)
        viewModel.updateTimer(myTimer)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initListener(){
        binding.settingsButton.setOnClickListener {
            (requireActivity() as MainActivity).openFragment(SettingsFragment())
        }
        binding.lockFrame.setOnTouchListener { _, _ ->
            showButtons()
            false
        }
        binding.lockButton.setOnClickListener { lockButton() }
    }

    private fun initView(){
        val arrayList = NightlightHelper.getNightlighters()
        binding.pager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)

                binding.bottomText.setText(arrayList[position].name)

                if (timerIsLoaded) {
                    val updatedTimer = myTimer.copy(
                        timerDuration = 9999,
                        timerStatus = false,
                    )
                    viewModel.updateTimer(updatedTimer)
                }
            }
        })

    }

    private fun initArrays(){
        bgColors = resources.getStringArray(R.array.bgColors)
        bgNlColors = resources.getStringArray(R.array.bgNlColors)
    }



}